### PR TITLE
feat: Landing Screen: Dynamic Type Support ACC-85

### PR DESCRIPTION
**Pull request by agisilaos**
<table>
<td>
  <a href="https://wearezeta.atlassian.net/browse/ACC-85" title="ACC-85" target="_blank">TaskACC-85</a>  [iOS] Landing Screen: Dynamic Type Support
  </td></table>
  <br />
 

----
#### PR Submission Checklist for internal contributors

- The **PR Title**
  - [x] conforms to the style of semantic commits messages¹ supported in Wire's Github Workflow²
  - [x] contains a reference JIRA issue number like `SQPIT-764`
  - [x] answers the question: _If merged, this PR will: ..._ ³

- The **PR Description**
  - [x] is free of optional paragraphs and you have filled the relevant parts to the best of your ability
----

# What's new in this PR?

With this PR we're trying to address a couple of issues and ultimately add dynamic type support for the landing screen in a more structured and concise way. As you can see there are a couple of new additions to the code base. First of all:

1. **Label.swift**. 

This a new class which is going to help us streamline most if not all of the labels in the codebase. In the init I replaced the titleLabelFont parameter with the size and weight so we can be more in sync with how we handle text across many parts of the app. What that class allow us to do is to be more safe in terms of the font and who's responsible to set it. 

2. **DynamicTypeCapable Protocol**.

That protocol will allow us to have a more streamlined way to handle Dynamic Type across many different Controls, components etc throughout the app. Every class which conforms to that protocol will have to implement a redrawFont method.

-- 

In addition to the changes I already mentioned I cleaned up things a bit. Removed unused methods from the LandingViewController.swift. Last but not least I've made some changes to the constraints so we can support Dynamic Type without any issues.

--

| iPhone SE - English | iPhone SE  - German | iPhone 8 - English | iPhone 8 - German |
|---|---|---|---|
| se - english | se - german | iPhone 8 - English | iPhone 8 - German |
----
##### References
1. https://sparkbox.com/foundry/semantic_commit_messages
1. https://github.com/wireapp/.github#usage
1. E.g. `feat(conversation-list): Sort conversations by most emojis in the title #SQPIT-764`.

### CODE REVIEW

File: Wire-iOS/Sources/Authentication/Landing/LandingViewController.swift [hunk 1]
```diff
@@ -99,8 +78,8 @@ final class LandingViewController: AuthenticationStepViewController {
         return label
     }()
 
-    private let subMessageLabel: UILabel = {
-        let label = UILabel(key: "landing.welcome_submessage".localized,
+    private let subMessageLabel: DynamicFontLabel = {
+        let label = DynamicFontLabel(text: L10n.Localizable.Landing.welcomeSubmessage,
```
Comment (johnxnguyen) [inline Wire-iOS/Sources/Authentication/Landing/LandingViewController.swift:82]: The indentation is off.

File: Wire-iOS/Sources/Authentication/Landing/LandingViewController.swift [hunk 2]
```diff
@@ -125,7 +104,7 @@ final class LandingViewController: AuthenticationStepViewController {
     }()
 
     private lazy var loginButton: Button = {
-        let button = Button(style: .full, variant: .light, titleLabelFont: .smallSemiboldFont)
+        let button = Button(style: .full, variant: .light, size: .small, weight: .semibold)
```
Comment (johnxnguyen) [inline Wire-iOS/Sources/Authentication/Landing/LandingViewController.swift:107]: `size` and `weight` are a bit confusing here because it's unclear what they're referring to (size of the button? Weight of the button?). I suggest to either rename them to `titleSize` and `titleWeight`, or add some helpers like this:

```swift
extension FontSpec {
  
    static var smallSemibold: Self {
        return self.init(.small, .semibold)
    }

}

let button = Button(style: .full, variant: .light, titleFont: .smallSemibold)

```

File: Wire-iOS/Sources/Authentication/Landing/LandingViewController.swift [hunk 3]
```diff
@@ -170,8 +148,8 @@ final class LandingViewController: AuthenticationStepViewController {
         return button
     }()
 
-    private let createAccoutInfoLabel: UILabel = {
-        let label = UILabel(key: "landing.create_account.infotitle".localized,
+    private let createAccoutInfoLabel: DynamicFontLabel = {
+        let label = DynamicFontLabel(text: L10n.Localizable.Landing.CreateAccount.infotitle,
```
Comment (johnxnguyen) [inline Wire-iOS/Sources/Authentication/Landing/LandingViewController.swift:152]: Indentation is off.

File: Wire-iOS/Sources/Authentication/Landing/LandingViewController.swift [hunk 4]
```diff
@@ -410,25 +389,18 @@ final class LandingViewController: AuthenticationStepViewController {
 
             // create an label
             createAccoutInfoLabelTopConstraint, // iPad
-            createAccoutInfoLabel.leadingAnchor.constraint(equalTo: view.leadingAnchor, constant: 24),
-            createAccoutInfoLabel.trailingAnchor.constraint(equalTo: view.trailingAnchor, constant: -24),
+            createAccoutInfoLabel.leadingAnchor.constraint(equalTo: view.leadingAnchor, constant: 5),
+            createAccoutInfoLabel.trailingAnchor.constraint(equalTo: view.trailingAnchor, constant: -5),
```
Comment (johnxnguyen) [inline Wire-iOS/Sources/Authentication/Landing/LandingViewController.swift:393]: Why are these needed?

File: Wire-iOS/Sources/Authentication/Landing/LandingViewController.swift [hunk 5]
```diff
@@ -40,28 +40,6 @@ final class LandingViewController: AuthenticationStepViewController {
...
-    static let buttonSubtitleAttribute: [NSAttributedString.Key: AnyObject] = {
-        let alignCenterStyle = NSMutableParagraphStyle()
-        alignCenterStyle.alignment = .center
-        alignCenterStyle.paragraphSpacingBefore = 4
-        alignCenterStyle.lineSpacing = 4
-
-        let lightFont = FontSpec(.normal, .light).font!
-
-        return [.foregroundColor: UIColor.Team.textColor, .paragraphStyle: alignCenterStyle, .font: lightFont]
-    }()
```
Comment (johnxnguyen) [inline Wire-iOS/Sources/Authentication/Landing/LandingViewController.swift:63]: Is this the dead code?

File: Wire-iOS/Sources/Components/Buttons/Button.swift [hunk 6]
```diff
@@ -27,8 +27,9 @@ enum ButtonStyle: Int {
     case emptyMonochrome
 }
 
-class Button: ButtonWithLargerHitArea {
+class Button: ButtonWithLargerHitArea, DynamicTypeCapable {
```
Comment (johnxnguyen) [inline Wire-iOS/Sources/Components/Buttons/Button.swift:30]: I suggest to create a `DynamicFontButton` subclass that contains only the `DynamicTypeCapable` conformance, then put this subclass in your inheritance hierarchy (e.g make `ButtonWithLargerHitArea` descend from `DynamicFontButton`. 

```swift
class DynamicFontButton: UIButton, DynamicTypeCapable {

    private let fontSpec: FontSpec

    init(fontSpec: FontSpec) {
        ...
    }

    func redrawFont() {
        ...
    }

}
```

File: Wire-iOS/Sources/Authentication/Landing/LandingViewController.swift (hunk 4 above)
Comment (agisilaos) [reply Wire-iOS/Sources/Authentication/Landing/LandingViewController.swift:393]: For Dynamic Type Purposes, for the text to fit when we increase the font size.

File: Wire-iOS/Sources/Authentication/Landing/LandingViewController.swift (hunk 2 above)
Comment (agisilaos) [reply Wire-iOS/Sources/Authentication/Landing/LandingViewController.swift:106]: I will create an extension for FontSpec and include every possible Font combination so we don't have to do it later.

File: Wire-iOS/Sources/UserInterface/Helpers/UIViewController+DynamicType.swift [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+
+}
+
+
+// MARK: - UIView Extension
+
+extension UIView {
+    /// We're going through each view which is a DynamicTypeCapable
+    /// and with we're calling the redrawFont method
+    func redrawAllfonts() {
```
Comment (KaterinaWire) [inline Wire-iOS/Sources/UserInterface/Helpers/UIViewController+DynamicType.swift:46]:
```suggestion
    func redrawAllFonts() {
```

**Comment by KaterinaWire** [comment]
@agisilaos Do you think we can provide any tests since the QA team will not be involved?

File: Wire-iOS/Sources/UserInterface/Helpers/UIViewController+DynamicType.swift (hunk 7 above)
Comment (agisilaos) [reply Wire-iOS/Sources/UserInterface/Helpers/UIViewController+DynamicType.swift:45]: Fixed with my latest commit.

**Comment by agisilaos** [comment]
@KaterinaWire and @johnxnguyen This PR is ready for  review.

File: Wire-iOS/Sources/Helpers/DynamicFontLabel.swift [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+    init(
+        text: String? = nil,
+        fontSpec: FontSpec,
+        color: ColorSchemeColor,
+        variant: ColorSchemeVariant = ColorScheme.default.variant
+    ) {
+        self.fontSpec = fontSpec
+        super.init(frame: .zero)
+
+        self.text = text.map { $0.localized }
```
Comment (johnxnguyen) [inline Wire-iOS/Sources/Helpers/DynamicFontLabel.swift:34]: Is `text` expected to be a localization key? Or should it already be localized? If it is expected to be a key, then the `text` argument should be renamed to something more appropriate.

File: Wire-iOS/Sources/Helpers/DynamicFontLabel.swift [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+        self.font = fontSpec.font
+        self.textColor = UIColor.from(scheme: color, variant: variant)
+        self.adjustsFontForContentSizeCategory = true
```
Comment (johnxnguyen) [inline Wire-iOS/Sources/Helpers/DynamicFontLabel.swift:37]: Is this necessary given that we don't use content size categories (yet...) ?

File: Wire-iOS/Sources/Helpers/DynamicFontLabel.swift [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+        self.translatesAutoresizingMaskIntoConstraints = false
+
+        self.fitIn(view: self)
```
Comment (johnxnguyen) [inline Wire-iOS/Sources/Helpers/DynamicFontLabel.swift:40]: This makes no sense, it seems like a left over from an old implementation where we had a nested label.

File: Wire-iOS/Sources/Helpers/DynamicFontLabel.swift [hunk 11]
```diff
@@ -0,0 +1,51 @@
...
+
+import UIKit
+
+class DynamicFontLabel: UILabel, DynamicTypeCapable {
+
+    private let fontSpec: FontSpec
+
+    init(
+        text: String? = nil,
+        fontSpec: FontSpec,
```
Comment (johnxnguyen) [inline Wire-iOS/Sources/Helpers/DynamicFontLabel.swift:27]: you may want to provide a sensible default value for the font spec, such as normal regular.

File: Wire-iOS/Sources/UserInterface/Helpers/FontSpec.swift [hunk 12]
```diff
@@ -0,0 +1,105 @@
...
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see http://www.gnu.org/licenses/.
+//
+
+extension FontSpec {
```
Comment (johnxnguyen) [inline Wire-iOS/Sources/UserInterface/Helpers/FontSpec.swift:19]: You might as well move the whole definition of `FontSpec` here, or rename the file to something like `FontSpec+Helpers`

File: Wire-iOS/Sources/UserInterface/Helpers/UIViewController+DynamicType.swift [hunk 13]
```diff
@@ -0,0 +1,60 @@
...
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see http://www.gnu.org/licenses/.
+//
+
+import UIKit
+
+// MARK: - UIViewController Extension
```
Comment (johnxnguyen) [inline Wire-iOS/Sources/UserInterface/Helpers/UIViewController+DynamicType.swift:21]: How about `// MARK - Helpers`?

File: Wire-iOS/Sources/UserInterface/Helpers/UIViewController+DynamicType.swift [hunk 14]
```diff
@@ -0,0 +1,60 @@
...
+
+extension UIViewController {
+
+    func redrawAllfonts() {
+        view.redrawAllFonts()
+    }
+
+}
+
+// MARK: - DynamicTypeCapable Protocol
```
Comment (johnxnguyen) [inline Wire-iOS/Sources/UserInterface/Helpers/UIViewController+DynamicType.swift:31]: I suggest to make `DynamicTypeCapable` the main thing in this file by moving it to the top and renaming the file 
`DynamicTypeCapable`.

File: Wire-iOS/Sources/UserInterface/Helpers/UIViewController+DynamicType.swift [hunk 15]
```diff
@@ -0,0 +1,60 @@
...
+
+/// Every view which conforms to that protocol has to implement a redrawFont method
```
Comment (johnxnguyen) [inline Wire-iOS/Sources/UserInterface/Helpers/UIViewController+DynamicType.swift:33]: This doc comment does not say anything about what the purpose of this protocol is, and in fact it should be clear from looking at it that the `redrawFont` method needs to be implemented.

I suggest something like "Objects conforming to this protocol opt in to react to changes of the preferred content size category".

File: Wire-iOS/Sources/UserInterface/Helpers/UIViewController+DynamicType.swift [hunk 16]
```diff
@@ -0,0 +1,60 @@
...
+protocol DynamicTypeCapable {
+    /// In this method we set the font for our views
+    func redrawFont()
```
Comment (johnxnguyen) [inline Wire-iOS/Sources/UserInterface/Helpers/UIViewController+DynamicType.swift:36]: Suggestion: "This method is called when the preferred content size category changes. Your implementation should update all of its fonts that are appropriately sized for the current content size category."

File: Wire-iOS/Sources/UserInterface/Helpers/UIViewController+DynamicType.swift [hunk 17]
```diff
@@ -0,0 +1,60 @@
...
+
+}
+
+// MARK: - UIView Extension
+
+extension UIView {
+    /// We're going through each view which is a DynamicTypeCapable
+    /// and with we're calling the redrawFont method
```
Comment (johnxnguyen) [inline Wire-iOS/Sources/UserInterface/Helpers/UIViewController+DynamicType.swift:44]: This comment is redundant as it does not add any new information that can be obtained by looking at the code.

File: Wire-iOS/Sources/Helpers/DynamicFontLabel.swift (hunk 8 above)
Comment (agisilaos) [reply Wire-iOS/Sources/Helpers/DynamicFontLabel.swift:34]: It's text (for now) to differentiate from the convenience init inside UILabel+Convenicence.swift file.

```swift
extension UILabel {
    convenience init(
        key: String? = nil,
        size: FontSize = .normal,
        weight: FontWeight = .regular,
        color: ColorSchemeColor,
        variant: ColorSchemeVariant = ColorScheme.default.variant) {
            self.init(frame: .zero)
            text = key.map { $0.localized }
            font = FontSpec(size, weight).font
            textColor = UIColor.from(scheme: color, variant: variant)
        }
```

File: Wire-iOS/Sources/UserInterface/Helpers/FontSpec.swift (hunk 12 above)
Comment (agisilaos) [reply Wire-iOS/Sources/UserInterface/Helpers/FontSpec.swift:19]: I don't want to do anything to that file right now as we're planning to move it to WireCommonComponents

File: Wire-iOS/Sources/UserInterface/Helpers/UIViewController+DynamicType.swift (hunk 14 above)
Comment (KaterinaWire) [reply Wire-iOS/Sources/UserInterface/Helpers/UIViewController+DynamicType.swift:31]: I agree that it's better to move this protocol on the top, since the name of the file is about `DynamicType`

File: Wire-iOS/Sources/UserInterface/Helpers/UIViewController+DynamicType.swift [hunk 18]
```diff
@@ -0,0 +1,60 @@
...
+    }
+
+}
+
+// MARK: - DynamicTypeCapable Protocol
+
+/// Objects conforming to this protocol opt in to react to changes of the preferred content size category
+protocol DynamicTypeCapable {
+    /// This method is called when the preferred content size category changes.
+    /// Your implementation should update all of its fonts that are appropriately sized for the current content size category."
```
Comment (KaterinaWire) [inline Wire-iOS/Sources/UserInterface/Helpers/UIViewController+DynamicType.swift:36]:
```suggestion
    /// Your implementation should update all of its fonts that are appropriately sized for the current content size category.
```

File: Wire-iOS/Sources/Components/Buttons/IconButton.swift [hunk 19]
```diff
@@ -59,15 +59,16 @@ class IconButton: ButtonWithLargerHitArea {
...
+        super.init(fontSpec: fontSpec)
 
         hitAreaPadding = CGSize(width: 20, height: 20)
     }
 
     convenience init(style: IconButtonStyle,
-                     variant: ColorSchemeVariant = ColorScheme.default.variant) {
-        self.init()
+                     variant: ColorSchemeVariant = ColorScheme.default.variant,
+                     fontSpec: FontSpec = .normalRegularFont) {
```
Comment (KaterinaWire) [inline Wire-iOS/Sources/Components/Buttons/IconButton.swift:70]: Why is default fontSpec here different from init()?

File: Wire-iOS/Sources/Helpers/DynamicFontLabel.swift [hunk 20]
```diff
@@ -0,0 +1,49 @@
...
+        variant: ColorSchemeVariant = ColorScheme.default.variant
+    ) {
+        self.fontSpec = fontSpec
+        super.init(frame: .zero)
+
+        self.text = text.map { $0.localized }
+        self.font = fontSpec.font
+        self.textColor = UIColor.from(scheme: color, variant: variant)
+        self.translatesAutoresizingMaskIntoConstraints = false
+
```
Comment (KaterinaWire) [inline Wire-iOS/Sources/Helpers/DynamicFontLabel.swift:38]:
```suggestion
```

File: Wire-iOS/Sources/Helpers/DynamicFontLabel.swift (hunk 8 above)
Comment (johnxnguyen) [reply Wire-iOS/Sources/Helpers/DynamicFontLabel.swift:34]: Then remove part where we localize it, since it should already be localized when passed into the initializer.

File: Wire-iOS/Sources/UserInterface/Helpers/UIViewController+DynamicType.swift (hunk 14 above)
Comment (agisilaos) [reply Wire-iOS/Sources/UserInterface/Helpers/UIViewController+DynamicType.swift:31]: I already moved that protocol to another file in the next PR.

File: Wire-iOS/Sources/Components/Buttons/IconButton.swift (hunk 19 above)
Comment (agisilaos) [reply Wire-iOS/Sources/Components/Buttons/IconButton.swift:70]: I don't think I had a specific reason in mind.

File: Wire-iOS/Sources/Helpers/DynamicFontLabel.swift (hunk 8 above)
Comment (agisilaos) [reply Wire-iOS/Sources/Helpers/DynamicFontLabel.swift]: Done!

File: Wire-iOS/Sources/UserInterface/Conversation/Content/Cells/ConfigurationMessageCell/ConversationMessageCell.swift [hunk 21]
```diff
@@ -105,7 +105,7 @@ extension ConversationMessageCell {
 
 protocol ConversationMessageCellDescription: AnyObject {
     /// The view that will be displayed for the cell.
-    associatedtype View: ConversationMessageCell & UIView
+    associatedtype View: ConversationMessageCell, UIView
```
[the remark this answers is not in the document]
Comment (agisilaos) [reply Wire-iOS/Sources/UserInterface/Conversation/Content/Cells/ConfigurationMessageCell/ConversationMessageCell.swift:108]: Yes!

File: Wire-iOS/Sources/Authentication/Landing/LandingViewController.swift [hunk 22]
```diff
@@ -170,12 +144,11 @@ final class LandingViewController: AuthenticationStepViewController {
...
     }()
 
-    private let createAccoutInfoLabel: UILabel = {
-        let label = UILabel(key: "landing.create_account.infotitle".localized,
-                            size: .small,
-                            weight: .regular,
-                            color: .landingScreen,
-                            variant: .light)
+    private let createAccoutInfoLabel: DynamicFontLabel = {
+        let label = DynamicFontLabel(text: L10n.Localizable.Landing.CreateAccount.infotitle,
```
[the remark this answers is not in the document]
Comment (agisilaos) [reply Wire-iOS/Sources/Authentication/Landing/LandingViewController.swift]: Done!

File: Wire-iOS/Sources/Authentication/Landing/LandingViewController.swift (hunk 2 above)
Comment (agisilaos) [reply Wire-iOS/Sources/Authentication/Landing/LandingViewController.swift]: Done!

**Review by johnxnguyen — APPROVED** [review]

File: Wire-iOS/Sources/UserInterface/Helpers/UIViewController+DynamicType.swift [hunk 23]
```diff
@@ -0,0 +1,60 @@
...
+// along with this program. If not, see http://www.gnu.org/licenses/.
+//
+
+import UIKit
+
+// MARK: - Helpers
+
+extension UIViewController {
+
+    func redrawAllfonts() {
```
Comment (KaterinaWire) [inline Wire-iOS/Sources/UserInterface/Helpers/UIViewController+DynamicType.swift:25]:
```suggestion
    func redrawAllFonts() {
```

**Review by KaterinaWire — APPROVED** [review]
